master: add table routing feature for cdc (#23211)
pingcap/docs: ticdc/ticdc-table-routing.md

---
title: TiCDC Table Routing
summary: Learn about table routing configuration in the TiCDC new architecture, including how to rewrite downstream database and table names using target-schema and target-table, how to handle DDL rewriting and routing conflicts, and how to troubleshoot common issues.
---

> # TiCDC Table Routing <span class="version-mark">Introduced in v8.5.7</span>

Author: ```suggestion
# TiCDC Table Routing <span class="version-mark">New in v8.5.7</span>
```

| The MQ Topic name still uses the upstream schema and table name. | Table routing does not change Topic or Partition dispatch rules. | If you need to change the Topic name, configure `topic` separately in `sink.dispatchers`. |
| The `CDC:ErrTableRoutingFailed` error is reported during DDL synchronization. | The DDL cannot be safely rewritten for table routing, or a schema-level DDL has ambiguity in the target schema. | Check the DDL type and routing rules. For schema-level DDL, make sure the same upstream schema can be parsed to only one target schema. |
| The Changefeed fails during runtime and reports the `CDC:ErrTableRouteConflict` error. | After a table is created or renamed, two different upstream tables are routed to the same downstream table. | Adjust the table routing rules or upstream DDL to ensure that, within a single Changefeed, each target table corresponds to only one upstream table being synchronized by the current Changefeed at any given time. |

## Related documentation

> - [TiCDC Changefeed command-line flags and configuration parameters](/ticdc/ticdc-changefeed-config.md)

Author: ```suggestion
- [CLI and Configuration Parameters of TiCDC Changefeeds](/ticdc/ticdc-changefeed-config.md)
```

> - [Changefeed event filter](/ticdc/ticdc-filter.md)

Author: ```suggestion
- [Changefeed Log Filters](/ticdc/ticdc-filter.md)
```

## Usage scenarios

You can configure table routing in the following scenarios:

- Replicating `sales.orders` to `archive.sales_orders`, or to a table that follows other downstream naming conventions.
- Replicating multiple source databases to the same target database while keeping target table names unique to distinguish the source database, for example, replicating `tenant_001.orders` to `tenant_mirror.tenant_001_orders`.
- Building changefeeds for migration, disaster recovery, archiving, or shadow traffic to avoid writing to downstream objects with the same names as upstream ones.
- Exposing stable database and table names to MQ consumer applications or storage services.

> **Note:**
>
> Table routing supports only a one-to-one mapping from an upstream table to a downstream target table. It does not support merging multiple upstream tables into the same downstream table.
> Table routing does not support splitting one upstream table into multiple downstream tables, or transforming row data content.

## Configure table routing

> Before configuring table routing, enable the TiCDC new architecture. For more information, see [`newarch`](/ticdc-server-config.md#newarch-new-in-v854-release1).

Author: ```suggestion
Before configuring table routing, enable the TiCDC new architecture. For more information, see [`newarch`](/ticdc/ticdc-server-config.md#newarch-new-in-v854-release1).
```

## Output behavior

| Sink | Behavior |
| :--- | :--- |
| MySQL sink | TiCDC writes DDL and DML statements to the routed target database and table. When the Redo feature is enabled, executing `redo apply` replays events to the routed target table. |
| Kafka sink and Pulsar sink | The values of the `payload` field in the protocol and the `query` field in DDL events use the routed target database and table names. The values of the `schema` and `table` fields in the encoding protocol also use the routed target database and table names. |
| Cloud storage sink | TiCDC generates the corresponding storage paths, schema files, table definition files, and data files based on the routed target database and table names. |

## DDL behavior

After you enable table routing, TiCDC rewrites DDL statements so that the database and table names in structured DDL metadata remain consistent with those in the SQL text.

For example, if the following rule is configured:

```toml
[[sink.dispatchers]]
matcher = ["sales.*"]
target-schema = "archive"
target-table = "{table}_routed"
```

> TiCDC rewrites the following upstream DDL:

Author: ```suggestion
For the following upstream DDL:
```

```sql
RENAME TABLE `sales`.`temp_table` TO `sales`.`renamed_table`;
```

> Into the following downstream DDL:

Author: ```suggestion
TiCDC rewrites it into the following downstream DDL:
```

> TiCDC rewrites it into the following downstream DDL:

Author: ````suggestion
For the following upstream DDL statement:

```sql
RENAME TABLE `sales`.`temp_table` TO `sales`.`renamed_table`;
```

TiCDC rewrites it into the following downstream DDL statement:
````
